common/type_system/TypeSystem.cpp

```diff
@@ -296,26 +297,23 @@ DerefInfo TypeSystem::get_deref_info(const TypeSpec& ts) const {
 
   if (ts.base_type() == "inline-array") {
     auto result_type = lookup_type(ts.get_single_arg());
-    auto result_structure_type = dynamic_cast<StructureType*>(result_type);
-    if (!result_structure_type || result_structure_type->is_dynamic()) {
+    if (auto result_structure_type = dynamic_cast<StructureType*>(result_type);
+        result_structure_type && result_structure_type->is_dynamic()) {
```

Comment:
I'd kind of prefer to leave this as it is, I think it is extra confusing that there are two ways to do the same thing, and it's hard to know if this really handles all cases.